### DISCUSSION TITLE
Should partial results on a details page work with graphcache if the details page query includes sub-fields?

### DISCUSSION THREAD

**Discussion by danoc** [urql-graphql/urql#1365] [Q&A] [answered]
Hello, 

Let's say I have a list view page (a list of todo items) and am trying to build a details view page that shows one item (a single todo). 

Using `@urql/exchange-graphcache` and providing a `schema`, I'm able to get partial results returning on the details page for users that first visit the list view. These partial results stop working, however, if the query on the details view page includes a sub-field such as `author`. 

Imagine that the query on the list view page looks like this:

```graphql
query GetTodos {
  todos {
      id
      name
  }
}
```

Imagine that the query on the details page looks like this:

```graphql
query GetTodoById($id: Int!) {
  todos_by_pk(id: $id) {
    id
    name
    authors {
      id
      name
    }
  }
}
```

The partial results only work if `authors { ... }` is removed from the query.

This is what the `cacheExchange` looks like:

```js
const cache = cacheExchange({
  schema,
  resolvers: {
    query_root: {
      todos_by_pk: (parent, args, cache, info) => {
        // This runs and works well
        return { __typename: "todos", id: args.id };
      },
      authors: (parent, args, cache, info) => {
        // This never gets run
        return { __typename: "authors", id: args.id };
      },
    },
  },
});
```

Should the details page be able to return partial results for the `todo` name, for example, even if it queries for `authors`?

Thank you for all of your effort on this!

**Comment by kitten** [comment, marked as answer]
That likely just means that the fields you're adding in your details query that aren't cached yet are marked as non-null

**Reply by danoc** [reply]
Ah, thanks. That's likely it.

What would be the downside for having it return the partial data regardless while it fetches the non-null data? Is the concern that the UI won't be built to handle fields that are required in the database but not yet available on the page?

**Reply by kitten** [reply]
Yes, typically what we're catering for are schema guarantees and UI guarantees. I'd see this as closely coupled as a schema always caters for a UI. There are multiple parts to this, but think of what happens if you got an error in part of your query. In such a case, GraphQL would also consult the API to find out how far up the schema it'd have to go to find out which parts of it can be nulled out.

In general if `authors` is either a Relay connection or a regular list then it may make sense to type the field itself a nullable if you're expecting that your UI will be able to handle it as a missing field.

This is the overall pattern that schema awareness follows for us. It allows Graphcache to independently make decisions about which parts are important or not (i.e. optional) so that it can still generate schema-compliant results while it's fetching more data.

**Reply by danoc** [reply]
Great, thank you Phil!

**Reply by marcel-happyfloat** [reply]
@kitten  Thanks for the explanation... I'm just switching over from apollo for my new project and I really struggled to find out why it does not resolve my entity from the cache.
I think the concept is pretty standard, the list view has limited data and the details view.... well ... has more details. So I really don't understand why it shouldn't display already all the information available. I can't just set half of my columns to null in the database... its required for other backend functionality. 
I don't think the UI/client(s) should have any influence on this. Is there any solution to my problem now or do I need switch back to apollo :(?

**Reply by kitten** [reply]
@happyfloat If you post a new thread with some query & code samples I can probably explain what's going on ✌️

**Reply by marcel-happyfloat** [reply]
@kitten Thank you very much for the fast reply. You already explained whats going on, my schema has `non-null` values, it works with nullable values. I think my problem is, that my graphql backend is auto-generated from the postgres schema ([hasura.io](https://hasura.io/)) and I can not change the non-null property if my postgres column is set to not null. The column must not be null for some backend logic, but the client/UI doesn't care about this.

I just noticed, that I can fake the missing cache data by just adding some random value in the link...
```js
content_by_activity: (data, args) => {
  return {
    __typename: 'content',
    id: args.activity_id,
    // adding this so I can add the missing cache field further down
    activity: {
      __typename: 'activity',
      id: args.activity_id,
      // missing non-null value in the cache, so we just add some value manually
      // and hope the network request will replace it later with the real value
      rec_participants: 666,
    },
  };
},
```

But this value never get's overwritten when I do 'cache-and-network'... so I'm out of solutions for now. Should I really create a new thread? I think it's about the same topic.

**Reply by austin43** [reply]
@kitten @happyfloat I ran into this same issue using Hasura.  It seems Hasura has a paradigm of schema generation tied to whether the database value is nullable.  This is at odds with graphcache assuming a field being "required" on a type means the cache is stale, when this could just mean you didn't want to query the field in the first place.  

Would love to hear if you were able to solve this, or if there is a workaround to set all type fields to optional if your UI is built to handle it with optional chaining.

**Reply by marcel-happyfloat** [reply]
@austin43 unfortunately not. I now believe that Hasura is doing it correctly. The graphql schema will never return null for the specified field, which is correct. The frontend client should absolutely have no influence on this. That's the main reason for Graphql... each client specified what it needs and it will get it accordingly.

I really don't understand how URQL handles the basic simple task of having a `ListView -> DetailView` and showing already partial cached data. It's just not working. Also I found a lot of re-render issues with the normalised graphcache... e.g. it will re-render my ListView because of some change in a different field.

**Reply by marcel-happyfloat** [reply]
So at the moment I dropped the Graphcache completely. Of course this hunts me on other places now and I will evaluate a new solution soon. Maybe I end up with Apollo again. It's also not perfect and bloated, but it handles all my use cases in one way or the other at least. (using it in another big project)

Please keep me informed if you find a solution. Thank you @austin43

**Reply by frederikhors** [reply]
> I really don't understand how URQL handles the basic simple task of having a `ListView -> DetailView` and showing already partial cached data. It's just not working.

It's working perfectly fine for us.

And it is the best GraphQL javascript library for now on the market. They did an amazing job. AMAZING.

You just need to study a little more to understand how things really work. And ask here - kindly and polite - because it's the only open source team I know that has always responded (in a very short time) with answers worthy of an academic doctorate.

**Reply by marcel-happyfloat** [reply]
@frederikhors Yes, the support is top notch, I also tried to get my answers here (https://github.com/FormidableLabs/urql/discussions/1834). Unfortunately I'm too dumb to understand the answer, so I gave up on it.

Maybe you can explain to me how this should work:

I have a ListView with Items which are requesting the `title` field. Then I have the DetailView, which is requesting an additional non-null field `xyz`. When I now navigate from the ListView to the DetailView, my`title` will be empty, because it's not resolving partial cached data because `xyz` is still missing. 

`xyz` is non-null in the schema, because the graphql api will always deliver a value for it (since it's also mandatory in the database). The client requests the field or does not request it. The client should not be able to influence the schema, because there will be multiple clients and every client will have different needs anyway.

This PR seems to maybe the solve the problem for me in the future:
https://github.com/FormidableLabs/urql/pull/2018

**Reply by frederikhors** [reply]
You need this: https://formidable.com/open-source/urql/docs/graphcache/schema-awareness/.

**Reply by frederikhors** [reply]
![image](https://user-images.githubusercontent.com/41120635/141110311-c1155b3a-a0c7-4dc6-9c3a-5085f085eb95.png)

**Reply by frederikhors** [reply]
Simpler than that is impossible.

**Reply by marcel-happyfloat** [reply]
This pic explains my problem perfectly. `details` is a `non-null` field in my graphql-schema and then it will not work like this.

**Reply by marcel-happyfloat** [reply]
Why should I make the field `nullable` if it will be never null when requested by the API? What If I have an additional ExtremeDetailsPage, where it is actually required to be non-null? How should I solve that?

**Reply by austin43** [reply]
@frederikhors 
> It's working perfectly fine for us.

Are you using hasura as well?  If so, would you be willing to help us get some insight on how you achieved this?

I think the concern here is that generated APIs that have a different concept of what a "required" field should mean will not work with urql's concept of ignoring cache redirects for entities that are missing said required field.  The only solution I can see at the moment is to somehow mark each field on the type as "optional" so urql doesn't throw out the redirect.

**Reply by kitten** [reply]
The short-answer is that basically the schema is stating what the database knows; however, we can only do so much in inferring from that. Hasura isn't wrong, Graphcache isn't wrong. The interpretation is just different. Hasura is very focused on exposing a database as a GraphQL schema and _not_ to make that GraphQL schema suitable for the client-side. That obviously creates a conflict, where it's convenient for Graphcache to assume that certain fields are nullable based on the schema, especially since it's common not to go "required-by-default".
Now, the long answer can be found here: https://github.com/FormidableLabs/urql/pull/2018
The GraphQL Working Group is aware of this and interested in finding a solution; that solution is the Nullability RFC, which hasn't shipped yet. The idea of it is to allow the client-side to override which fields may be required or nullable, which perfectly suits your case and the partial data case. Apart from that `@defer` is obviously another option, which is also supported by urql and Graphcache, but presumably not by Hasura yet, since it hasn't shipped yet.

**Reply by austin43** [reply]
@kitten This makes sense to me.  If I'm reading you correctly, in its current state, Hasura users cannot use partial cache redirects until nullability comes to the spec or @defer comes to Hasura.  Not a dealbreaker at all for us, but may be for someone reading this.

Thank you for the answer and for the great work on urql!

**Reply by marcel-happyfloat** [reply]
@kitten Thanks for confirming and explaining. I still don't understand how a Graphql Schema should be made suitable for one specific client. Even for the same client different views can have different requirements. This goes extremely against my intuition for what GraphQL was made for, e.g. be client independent. So I'm really looking forward for the `Nullability RFC`. Thanks!

**Reply by marcel-happyfloat** [reply]
Very interesting read about the new "@required" directive from relay: https://relay.dev/docs/guides/required-directive/ . Especially this part seems related to the core of the problem here: https://relay.dev/docs/guides/required-directive/#why-not-implement-this-at-the-schemaserver-level

**Reply by ghost** [reply]
what about providing a full example of this in react?
